# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv](https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive-beta.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv")
data.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [3]:
# counting how many values we have in each category...
data['spam'].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

In this exercise we are applying two different models, logistic regression and random forest, to a spam email dataset to see which does a better job of learning to identify spam emails. Logistic regression is a linear model that works well with binary classification tasks, such as spam detection. it works well when the relationship between the features of the data set and the target variable are easily approximated by a straight line or plane. It is computationally efficient, which makes it a good choice for tassk with a large number of features and a moderately sized data set. It doesn't do well with capturing complex, non-linear relationships between the features and the target.

Random Forest is an ensemble learning method based on decision trees. it's a highly flexible method that is better suited to capturing complex relationships between features and a target variable. It handles noisy data well, however, it can be computationally more expensive than compared to logistic regression.

Comparing these two methods, I would expect that logistic regression would do well capturing the relationship between features, such as keywords, and the target spam/"ham" classification. However, I would expect that, even when they do well, most linear models are an oversimplification of the relationship between a datasets' features and the clasification. Linear relationships will model the dominant relationships in the dataset, but there is probably additional information that could still be captured with a different model that might give us better precision and recall for the target classification.

I suspect that while spam email datasets generally do have strong correlations between features and the target variable that have traditionally made them strong candidates for linear statistical techniques such logistic regression, there is enough underlying complexity in the relationship between the various features and the target that a non-linear model might very well do a better job of capturing that relationship and have better precision and recall.


## Split the Data into Training and Testing Sets

In [4]:
# Create the labels set `y` and features DataFrame `X`
X = data.copy()
X = X.drop(columns='spam')
y = data['spam']

In [5]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [6]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X,y)
X_train

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4031            0.00               0.00           0.00           0.0   
974             0.90               0.00           0.00           0.0   
4390            0.00               0.00           1.72           0.0   
728             0.00               0.00           0.00           0.0   
2041            0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
768             0.00               0.00           0.18           0.0   
3353            0.00               0.00           0.00           0.0   
78              0.00               0.41           0.20           0.0   
4048            0.26               0.00           0.52           0.0   
691             0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4031           0.32            0.00              0.00                0.00   
974            0.00            0.00              0.90                0.00   
4390           0.00            0.00              0.00                0.00   
728            0.00            0.00              0.00                0.00   
2041           0.00            0.00              0.00                0.00   
...             ...             ...               ...                 ...   
768            0.55            0.37              0.18                0.18   
3353           2.38            0.00              0.00                0.00   
78             1.67            0.20              0.20                0.00   
4048           0.52            0.00              0.00                0.00   
691            0.00            0.00              0.00                0.74   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4031              0.0            0.00  ...                   0.0        0.000   
974               0.0            1.80  ...                   0.0        0.000   
4390              0.0            0.00  ...                   0.0        0.000   
728               0.0            0.00  ...                   0.0        0.000   
2041              0.0            0.00  ...                   0.0        0.000   
...               ...             ...  ...                   ...          ...   
768               0.0            0.18  ...                   0.0        0.000   
3353              0.0            0.00  ...                   0.0        0.000   
78                0.0            1.04  ...                   0.2        0.000   
4048              0.0            0.26  ...                   0.0        0.047   
691               0.0            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4031        0.094        0.047        0.094        0.000        0.000   
974         0.000        0.000        0.391        0.130        0.260   
4390        0.303        0.000        0.910        0.000        0.000   
728         0.000        0.000        0.000        0.000        0.000   
2041        0.163        0.000        0.000        0.000        0.000   
...           ...          ...          ...          ...          ...   
768         0.081        0.000        0.000        0.027        1.625   
3353        0.000        0.000        0.000        0.000        0.000   
78          0.000        0.000        0.132        0.000        0.000   
4048        0.047        0.000        0.047        0.000        0.000   
691         0.000        0.000        0.109        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
4031                       1.919                          13   
974                        2.150                           7   
4390                       2.171                           9   
728                        1.055                           3   
2041                       3.280              

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [7]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [8]:
# Fit the Standard Scaler with the training data
X_scaler = scaler.fit(X_train)

In [9]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score.

In [10]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
logistic_regression_model = LogisticRegression(random_state=1,max_iter=2000)
logistic_regression_model.fit(X_train_scaled, y_train)
print(f"Training Data Score: {logistic_regression_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {logistic_regression_model.score(X_test_scaled, y_test)}")

Training Data Score: 0.9315942028985508
Testing Data Score: 0.9331016507384883


In [11]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_regression_model.predict(X_test_scaled)
# Review the predictions
testing_df = pd.DataFrame({"Prediction": testing_predictions, "Actual":y_test})
testing_df

Prediction  Actual
2867           0       0
1975           0       0
2523           0       0
4386           1       0
4145           0       0
...          ...     ...
167            1       1
4506           0       0
1284           0       1
2185           0       0
410            1       1

[1151 rows x 2 columns]

In [12]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
lr_score = accuracy_score(y_test,testing_predictions)
print(f"Logistic Regression Accuracy : {lr_score}")

Logistic Regression Accuracy : 0.9331016507384883


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score.

In [13]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=128, random_state=78)
rf_model = rf_model.fit(X_train_scaled, y_train)
print(f"Random Forest Training Data Score: {rf_model.score(X_train_scaled, y_train)}")

Random Forest Training Data Score: 0.9997101449275362


In [14]:
# Make and save testing predictions with the saved logistic regression model using the test data
# I assume this is supposed to say make predictions using the random forest model
rf_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_testing_df = pd.DataFrame({"Prediction": rf_predictions, "Actual":y_test})
rf_testing_df


Prediction  Actual
2867           0       0
1975           0       0
2523           0       0
4386           0       0
4145           0       0
...          ...     ...
167            1       1
4506           0       0
1284           1       1
2185           0       0
410            1       1

[1151 rows x 2 columns]

In [15]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_score = accuracy_score(y_test, rf_predictions)
print(f"Random Forest Accuracy Score : {rf_score}")

Random Forest Accuracy Score : 0.9635099913119027


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

From what I'm seeing, the Random Forest Accuracy Score is 95% and the Accuracy Score for Logistic Regression is 93%. This would bear out my assertion that while the linear models work well with spam email data sets because there are are strong linear relationships between features and the binary classification that we are looking for, this still represents an oversimplification and non-linear models can capture more of this complexity and give us higher accuracy.